runtime/lua/vim/pack.lua

```diff
@@ -0,0 +1,653 @@
+--- @param counter function
+local function clone_or_pull(pkg, counter)
+  if Filter.to_update(pkg) then
+    pull(pkg, counter)
+  elseif Filter.to_install(pkg) then
+    clone(pkg, counter)
+  end
+end
+
+local function process_build_queue()
+  local failed = {}
+
+  local after = function(pkg, ok)
+    report(pkg.name, Messages.build, ok)
+    if not ok then
+      table.insert(failed, pkg)
+    end
+  end
+
+  for _, pkg in ipairs(BuildQueue) do
+    local t = type(pkg.build)
+    if t == 'function' then
+      ---@diagnostic disable-next-line: param-type-mismatch
+      local ok = pcall(pkg.build)
+      after(pkg, ok)
+    elseif t == 'string' and vim.startswith(pkg.build, ':') then
+      ---@diagnostic disable-next-line: param-type-mismatch
+      local ok = pcall(vim.cmd, pkg.build)
+      after(pkg, ok)
+    elseif t == 'string' then
+      local args = vim.split(pkg.build, '%s', { trimempty = true })
+      vim.system(
+        args,
+        { cwd = pkg.dir },
+        vim.schedule_wrap(function(obj)
+          after(pkg, obj.code == 0)
+        end)
+      )
+    end
+  end
+
+  BuildQueue = failed
+end
+
+--- @param pkg vim.pack.Package
+local function reclone(pkg)
+  local ok = pcall(vim.fs.rm, pkg.dir, { recursive = true })
+  if ok then
+    clone(pkg, function() end)
+  end
+end
+
+--- @param conflict pack.Conflict
+local function resolve(conflict)
+  reclone(conflict.curr)
+end
+
+--- @param pkg string|vim.pack.PackageSpec
+--- @return vim.pack.Package
+local function register(pkg)
+  if type(pkg) == 'string' then
+    pkg = { pkg }
+  end
+
+  local url = (pkg[1]:match('^https?://') and pkg[1]) -- [1] is a URL
+    or string.format(Config.url_format, pkg[1]) -- [1] is a repository name
+
+  local name = pkg.as or url:gsub('%.git$', ''):match('/([%w-_.]+)$') -- Infer name from `url`
+  if not name then
+    error('Failed to parse ' .. vim.inspect(pkg))
+  end
+
+  local dir = vim.fs.joinpath(Path.packs, name)
+  local ok, hash = pcall(get_git_hash, dir)
+  hash = ok and hash or ''
+
+  return {
+    branch = pkg.branch,
+    build = pkg.build,
+    dir = dir,
+    hash = hash,
+    name = name,
+    opt = pkg.opt,
+    pin = pkg.pin,
+    status = uv.fs_stat(dir) and M.status.INSTALLED or M.status.TO_INSTALL,
+    url = url,
+  }
+end
+
+--- @param pkg vim.pack.Package
+--- @param counter function
+local function remove(pkg, counter)
+  local ok = pcall(vim.fs.rm, pkg.dir, { recursive = true })
+  counter(pkg.name, Messages.remove, ok and 'ok' or 'err')
+  if not ok then
+    return
+  end
+  pkg.status = M.status.REMOVED
+  Packages[pkg.name] = pkg
+end
+
+--- @nodoc
+--- @class pack.Conflict
+--- @field prev vim.pack.Package
+--- @field curr vim.pack.Package
+---
+--- @return pack.Conflict[]
+local function calculate_conflicts()
+  local conflicts = {}
+  for name, lock in pairs(Lock) do
+    local pkg = Packages[name]
+    if pkg and Filter.not_removed(lock) and (lock.branch ~= pkg.branch or lock.url ~= lock.url) then
```

Comment:
```suggestion
    if pkg and Filter.not_removed(lock) and (lock.branch ~= pkg.branch or lock.url ~= pkg.url) then
```

While its other comments were nonsense, the review bot that invaded us was right about this at least.